corehq/apps/couch_sql_migration/asyncforms.py

```diff
@@ -0,0 +1,329 @@
+
+        Returns :boolean: True if it acquired the lock, False if it was added to queue
+        """
+        if not lock_ids:
+            self._add_item(lock_ids, queue_obj, to_queue=False)
+            return True
+        if self._check_lock(lock_ids):  # if it's currently locked, it can't acquire the lock
+            self._add_item(lock_ids, queue_obj)
+            return False
+        for lock_id in lock_ids:  # if other objs are waiting for the same locks, it has to wait
+            queue = self.queue_by_lock_id[lock_id]
+            if queue:
+                self._add_item(lock_ids, queue_obj)
+                return False
+        self._add_item(lock_ids, queue_obj, to_queue=False)
+        self._set_lock(lock_ids)
+        return True
+
+    def get_next(self):
+        """Returns the next object and lock ids that can be processed
+
+        Iterates through the first object in each queue, then checks
+        that that object is the first in every lock queue it is in.
+
+        :returns: A tuple: `(<queue_obj>, <lock_ids>)`; `(None, None)`
+        if nothing can acquire the lock currently.
+        """
+        def is_first(queue_id, lock_id):
+            return queue_by_lock_id[lock_id][0] == queue_id
+
+        queue_by_lock_id = self.queue_by_lock_id
+        lock_ids_by_queue_id = self.lock_ids_by_queue_id
+        for queue in six.itervalues(queue_by_lock_id):
+            if not queue:
+                continue
+            queue_id = queue[0]
+            lock_ids = lock_ids_by_queue_id[queue_id]
+            if all(is_first(queue_id, x) for x in lock_ids) and self._set_lock(lock_ids):
+                return self._pop_queue_obj(queue_id), lock_ids
+        return None, None
+
+    def has_next(self):
+        """ Makes sure there are still objects in the queue
+
+        Returns :boolean: True if there are objs left, False if not
+        """
+        for queue in six.itervalues(self.queue_by_lock_id):
+            if queue:
+                return True
+        return False
+
+    def release_lock_for_queue_obj(self, queue_obj):
+        """ Releases all locks for an object in the queue
+
+        :queue_obj obj: An object of the type in the queues
+
+        At some point in the future it might raise an exception if it trys
+        releasing a lock that isn't held
+        """
+        queue_obj_id = self.get_queue_obj_id(queue_obj)
+        lock_ids = self.lock_ids_by_queue_id.pop(queue_obj_id, None)
+        if lock_ids:
+            self._release_lock(lock_ids)
+            return True
+        return False
+
+    @property
+    def remaining_items(self):
+        return len(self.queue_objs_by_queue_id)
+
+    @property
+    def full(self):
+        if self.max_size == -1:
+            return False
+        return self.remaining_items >= self.max_size
+
+    def _add_item(self, lock_ids, queue_obj, to_queue=True):
+        """
+        :to_queue boolean: adds object to queues if True, just to lock tracking if not
```

Comment:
nit: `to lock tracking` -> `do lock tracking`

Reply:
```suggestion
        :to_queue boolean: adds object to queues if True, just do lock tracking if not
```